pkg/ebpf/common/dns_request_transform.go

```diff
@@ -0,0 +1,135 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ebpfcommon
+
+import (
+	"net"
+	"strings"
+	"unsafe"
+
+	"golang.org/x/net/dns/dnsmessage"
+
+	"go.opentelemetry.io/otel/trace"
+
+	"go.opentelemetry.io/obi/pkg/appolly/app/request"
+	"go.opentelemetry.io/obi/pkg/ebpf/common/dnsparser"
+	"go.opentelemetry.io/obi/pkg/internal/ebpf/ringbuf"
+	"go.opentelemetry.io/obi/pkg/pipe/msg"
+)
+
+func dnsEventExpireHandler(spansChan *msg.Queue[[]request.Span], filter ServiceFilter) func(key dnsparser.DNSId, span *request.Span) {
+	return func(_ dnsparser.DNSId, span *request.Span) {
+		// final status is -1, which means we never received a response
+		if span.Status == -1 {
+			span.Status = int(dnsparser.RCodeRefused)
+			spansChan.Send(filter.Filter([]request.Span{*span}))
+		}
+	}
+}
+
+func ReadDNSEventIntoSpan(parseCtx *EBPFParseContext, record *ringbuf.Record) (request.Span, bool, error) {
```

Comment:
I think this method can be private, so we keep the public API as smallest as possible